### PR TITLE
JM - implement optional remember_me to persist authentication

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

Please split this task up into three separate PRs. The second, and third should
depend on the preceding ones.

__Acceptance criteria PR1__:
- [x] Users can sign up.

__Acceptance criteria PR2__:
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [x] Logged in users should not be able to register, nor log in again.

Reviewer: grahamotte

__Acceptance criteria PR3__:
- [x] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [x] Multiple browsers can concurrently maintain the remembered status for the
  same user.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser. Please be sure to completely
close the browser (`cmd` + `q`) when testing this functionality. Closing the
window (`cmd` + `w`, or clicking the red circle) on Mac OS does not close the
browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by jmangel**
Closes #12

### CODE REVIEW

File: app/models/forms/user_form.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module Forms
+  class UserForm
```
Comment (bboe) [inline app/models/forms/user_form.rb]: Awesome

File: test/models/forms/user_form_test.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  class UserFormTest < ActiveSupport::TestCase
+    test 'initialize user form' do
+      user_form = UserForm.new(name: 'Example User', email: 'user@example.com',
+                               password: 'foobar', password_confirmation: 'foobar', remember_me: true)
+
+      assert_equal 'Example User', user_form.name
+      assert_equal 'user@example.com', user_form.email
+      assert_equal 'foobar', user_form.password
+      assert_equal 'foobar', user_form.password_confirmation
+      assert user_form.remember_me
```
Comment (bboe) [inline test/models/forms/user_form_test.rb]: Consider using `assert_predicate`.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -6,15 +6,18 @@ def show
...
   def create
-    @new_user = User.new(user_params)
-    if @new_user.save
-      log_in @new_user
-      redirect_to @new_user, flash: { success: 'You have successfully signed up' }
+    user = User.new(user_params)
+    if user.save
+      login user
     else
+      @new_user = Forms::UserForm.new(user_form_params)
```
Comment (bboe) [inline app/controllers/users_controller.rb]: In this case, what do you get from the form object that you don't get from the User object?

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -3,8 +3,24 @@ def log_in(user)
...
+    cookies.delete(:user_id)
+  end
+
   def current_user
-    @current_user ||= User.find_by(id: session[:user_id])
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: I'm curious in what cases user might not be found here? Is that something that should be an exception (very unlikely), or something that should actually be handled?

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (jmangel) [reply app/controllers/users_controller.rb]: :remember_me

File: app/helpers/sessions_helper.rb (hunk 4 above)
Comment (jmangel) [reply app/helpers/sessions_helper.rb]: If the user is not found, current_user will return `nil`, and `logged_in?` will return false. If this happened, for example, in the case where a user is deleted from the db while that user is logged in, that user will simply be logged out from then on. I think this is acceptable function, how about you? @bboe  
Also, `session[:user_id]` and `cookies.signed[:user_id]` only get set by `log_in` and `remember`, respectively, which both get passed a user that is found with `User.find_by(email: params[:user][:email].downcase)` in the sessions_controller if `user` is not nil. So I think it should be found as long as the user is not deleted after login.

**Pull request merged**